File: rust_exercises/leetcode/Problem736/Solution.java
class Solution {
    public int evaluate(String expression) {
        Map<Integer, Integer> paren = new HashMap<>();
        Stack<Integer> temp = new Stack<>();
        
        for (int i = 0; i < expression.length(); i += 1) {
            char c = expression.charAt(i);
            
            if (c == '(') {
                temp.push(i);
            } else if (c == ')') {
                paren.put(temp.pop(), i);
            }
        }
        
        return evaluate(expression, 0, expression.length() - 1, new HashMap<>(), paren);
    }
    
    private int evaluate(String expression, int l, int r, Map<String, Stack<Integer>> values, Map<Integer, Integer> paren) {
        if (expression.charAt(l) == '(') {
            return evaluate(expression, l + 1, r - 1, values, paren);
        }
        
        if (r - l >= 3 && expression.substring(l, l + 4).equals("let ")) {
            Stack<String> new_assigned = new Stack<>();
            int pos = l + 4;
            
            while (true) {
                
                int vr = get_right(expression, pos, r, paren);
                if (vr == r) {
                    break;
                }
                String v = expression.substring(pos, vr + 1);
                pos = vr + 2;
                
                int er = get_right(expression, pos, r, paren);
                int e = evaluate(expression, pos, er, values, paren);
                pos = er + 2;
                
                new_assigned.push(v);
                if (!values.containsKey(v)) {
                    values.put(v, new Stack<>());
                }
                values.get(v).push(e);
            }
            
            int ret = evaluate(expression, pos, r, values, paren);
            while (!new_assigned.isEmpty()) {
                String v = new_assigned.pop();
                values.get(v).pop();
                if (values.get(v).isEmpty()) {
                    values.remove(v);
                }
            }
            return ret;
        }
        
        if (r - l >= 3 && expression.substring(l, l + 4).equals("add ")) {
            int pos = l + 4;
            
            int e1r = get_right(expression, pos, r, paren);
            int e1 = evaluate(expression, pos, e1r, values, paren);
            pos = e1r + 2;
            
            int e2 = evaluate(expression, pos, r, values, paren);
            
            return e1 + e2;
        }
        
        if (r - l >= 4 && expression.substring(l, l + 5).equals("mult ")) {
            int pos = l + 5;
            
            int e1r = get_right(expression, pos, r, paren);
            int e1 = evaluate(expression, pos, e1r, values, paren);
            pos = e1r + 2;
            
            int e2 = evaluate(expression, pos, r, values, paren);
            
            return e1 * e2;
        }
        
        
        if (Character.isLowerCase(expression.charAt(l))) {
            return values.get(expression.substring(l, r + 1)).peek();
        }
        
        return Integer.parseInt(expression.substring(l, r + 1));
    }
    
    private int get_right(String expression, int l, int r, Map<Integer, Integer> paren) {
        if (expression.charAt(l) == '(') {
            return paren.get(l);
        }
        
        int right = l;
        while (right <= r && expression.charAt(right) != ' ') {
            right += 1;
        }
        return right - 1;
    }
}

File: rust_exercises/leetcode/Problem736/solution.rs
use std::collections::HashMap;
use std::cell::RefCell;

impl Solution {
    pub fn evaluate(expression: String) -> i32 {
        let mut paren = HashMap::new();
        let mut temp = vec![];
        
        for (i, c) in expression.chars().enumerate() {
            if c == '(' {
                temp.push(i);
            } else if c == ')' {
                paren.insert(temp.pop().unwrap(), i);
            }
        }
        
        let mut values = HashMap::new();
        
        eval(&expression, 0, expression.len() - 1, &mut values, &paren)
    }
}

fn eval(expr: &String, l: usize, r: usize, values: &mut HashMap<String, RefCell<Vec<i32>>>, paren: &HashMap<usize, usize>) -> i32 {
    if &expr[l..(l + 1)] == "(" {
        return eval(expr, l + 1, r - 1, values, paren);
    }
    
    if r - l >= 3 && &expr[l..(l + 4)] == "let " {
        let mut new_assigned = vec![];
        let mut pos = l + 4;
        
        loop {
            let vr = get_right(expr, pos, r, paren);
            if vr == r {
                break;
            }
            let v = String::from(&expr[pos..(vr + 1)]);
            pos = vr + 2;
            
            let er = get_right(expr, pos, r, paren);
            let e = eval(expr, pos, er, values, paren);
            pos = er + 2;
            
            new_assigned.push(v.clone());
            if !values.contains_key(&v) {
                values.insert(v.clone(), RefCell::new(vec![]));
            }
            values.get(&v).unwrap().borrow_mut().push(e);
        }
        
        let ret = eval(expr, pos, r, values, paren);
        while new_assigned.len() > 0 {
            let v = new_assigned.pop().unwrap();
            values.get(&v).unwrap().borrow_mut().pop();
            if values.get(&v).unwrap().borrow().len() == 0 {
                values.remove(&v);
            }
        }
        return ret;
    }
    
    if r - l >= 3 && &expr[l..(l + 4)] == "add " {
        let mut pos = l + 4;
        
        let e1r = get_right(expr, pos, r, paren);
        let e1 = eval(expr, pos, e1r, values, paren);
        pos = e1r + 2;
        
        let e2 = eval(expr, pos, r, values, paren);
        
        return e1 + e2;
    }
    
    if r - l >= 4 && &expr[l..(l + 5)] == "mult " {
        let mut pos = l + 5;
        
        let e1r = get_right(expr, pos, r, paren);
        let e1 = eval(expr, pos, e1r, values, paren);
        pos = e1r + 2;
        
        let e2 = eval(expr, pos, r, values, paren);
        
        return e1 * e2;
    }
    
    let c = expr.as_bytes()[l] as char;
    if (c.is_lowercase()) {
        let temp = values.get(&expr[l..(r + 1)]).unwrap().borrow();
        return temp[temp.len() - 1];
    }
    
    String::from(&expr[l..(r + 1)]).parse().unwrap()
}


fn get_right(expr: &String, l: usize, r: usize, paren: &HashMap<usize, usize>) -> usize {
    if &expr[l..(l + 1)] == "(" {
        return *paren.get(&l).unwrap();
    }
    
    let mut right = l;
    while right <= r && &expr[right..(right + 1)] != " " {
        right += 1;
    }
    
    right - 1
}
